### ISSUE TITLE
Dynamic remoteEntry.js URL plugin

### ISSUE THREAD

**Issue by matthewma7** [module-federation/module-federation-examples#566]
Hi, I wrote a plugin to help solve the following two needs we had for module federation. I got the permission to open source it to help the community of module federation. 
- Dynamic URL, have the ability to define the URL at runtime instead of hard code at build time.  
- Cache invalidation. 
Append random query string to the end of the remoteentry.js to have better control of cache invalidation. For example, the `getRandomString()` below will generate the same random string in a 5 minutes window based on current epoch time. 

Usage,
```javascript
// webpack.config.js

plugins: [
    new ModuleFederationPlugin({
        name: 'my-app',
        remotes: {
            'my-remote-1': 'my-remote-1@[window.remote-1-domain]/remoteEntry.js?[getRandomString()]',
            ...
        },
        shared: {...}
    }),
    new ExternalTemplateRemotesPlugin(), //no parameter,
    ...otherPlugins,
]
```
Basically, this plugin modifies webpack generated code and transform the template string into a javascript expression, which it is able to read values from the global javascript context. Since module federation uses webpack external module, the plugin is applied to external module, not solely for module federation.

Here is the gist of the file. I use a gist instead of pasting here to have potential fixes.  
https://gist.github.com/matthewma7/915b28846ad99d582e6b4ddea6dfc309

**Comment by davekiss** [comment]
@matthewma7 this looks great, do you have an example of where to put the plugin so it can be included in the webpack file? maybe it makes sense as an npm package?

**Comment by matthewma7** [comment]
@davekiss At this moment, I am thinking people would just copy and paste the gist file into your repo, until I sense it's useful for many people, then I would publish to the npm. 🙂

**Comment by jacob-ebey** [comment]
Thank you for sharing

**Comment by Guriqbal-Singh-Alida** [comment]
@matthewma7 nice work. Any plans to release it to npm?

**Comment by Guriqbal-Singh-Alida** [comment]
for others, usage is as follows:
First initialize the plugin as in https://github.com/module-federation/module-federation-examples/issues/566#issue-785273439

then in your  loading file call it like
```
import bootstrap from './bootstrap';
import config from './config';

window.__my_var__ = config.urlString;

OR

//for async
config.getRemoteAsync().then((x) => {
  window.__my_var__ = x.urlString;
  bootstrap(() => {});
});
```

Thanks again to @matthewma7 for sharing the nice and smart solution

**Comment by ksrb** [comment]
For those who are still confused by the provided examples this is what you have to do:

**Host (app1) webpack.config**
```js
const config = {
   ...otherConfigs
  plugins: [
    new ModuleFederationPlugin({
      name: "app1",
      remotes: {
        app2: "app2@[window.app2Url]/remoteEntry.js"
      }
    }).
    new ExternalTemplateRemotesPlugin(),
  ]
}
```

**Host (app1) source somewhere before loading main entry file**
```js
window.app2Url = "//localhost:3002"; // Whatever the url/logic to determine your remote module is

import("./bootstrap");
```

Working example is also available in this PR: https://github.com/module-federation/module-federation-examples/pull/557.

**Update:** the PR was merged a example can be found under [advanced-api/dynamic-remotes-synchronous-imports](https://github.com/module-federation/module-federation-examples/tree/d5cf265c2d4fd040797cbae806badd8267ad5b8f/advanced-api/dynamic-remotes-synchronous-imports)

**Comment by XiaofengXie16** [comment]
Any updates on publishing this package to NPM?

**Comment by ScriptedAlchemy** [comment]
Ill work on creating an npm package under module federation tomorrow @skyrex-mark 

thanks @ksrb

**Comment by XiaofengXie16** [comment]
> Ill work on creating an npm package under module federation tomorrow @skyrex-mark
> 
> thanks @ksrb

It seems like the repo is created but it is not published to npm yet.

**Comment by ScriptedAlchemy** [comment]
Plugin has been published to npm

**Comment by arshita04** [comment]
> Plugin has been published to npm

With what name has it been published? Couldn't find it in npm. Also, after using this plugin are we loading the remotes dynamically? the way we are doing in the dynamic-remotes example.

**Comment by gothy** [comment]
@arshita04 i believe this is it https://www.npmjs.com/package/external-remotes-plugin

**Comment by vospascal** [comment]
idea cant there be like a <link rel="dependencies" href="dependencies.json"> like manifest.json and so your main html which knows his context can load the proper stuff something that could be build from inside webpack but also be managed externaly ? 
you could do dependencies per scope/page or just flat structure. since its there on inital load it can be used easly.

**Comment by jacob-ebey** [comment]
@vospascal It already exists in the form of: https://github.com/jacob-ebey/webpack-federated-stats-plugin

**Comment by vospascal** [comment]
> @vospascal It already exists in the form of: https://github.com/jacob-ebey/webpack-federated-stats-plugin

Nice didn’t know that thnx

**Comment by AdamWold** [comment]
Should this work with an Angular application? The webpack config seems to be processed at compile time and at runtime I'm getting an error because the remoteEntry url wasn't updated: 
![image](https://user-images.githubusercontent.com/23104158/173121020-91a3f13c-6192-4718-9402-7fd073f341f2.png)
![image](https://user-images.githubusercontent.com/23104158/173121106-b32f1201-3bdf-40bf-ace9-602ceae65b95.png)
![image](https://user-images.githubusercontent.com/23104158/173121323-987e3402-b190-4d17-a3ac-1298d7c96e02.png)

**Comment by AdamWold** [comment]
Seems my issue may be that this was written for Webpack < 5?

**Comment by nkalinov** [comment]
Hey guys, thank you all your work here! 
Is there a way to provide remote URL fallback? 
Ie: try to load module "foo" from URL-1, if it fails (404 for example), fallback to loading the from URL-2 ?
Trying to make local development easier by loading some modules from a CDN env if it is not running on localhost...

**Comment by XiaofengXie16** [comment]
> Hey guys, thank you all your work here! Is there a way to provide remote URL fallback? Ie: try to load module "foo" from URL-1, if it fails (404 for example), fallback to loading the from URL-2 ? Trying to make local development easier by loading some modules from a CDN env if it is not running on localhost...

I don't think there is built-in re-try logic for this plugin. However, this maybe achievable with the `promise new promise` syntax. You can basically define whatever logic you want with that syntax.
reference: https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes

**Comment by ChristopherHButler** [comment]
I am a bit late to the party but in my own Net project I have observed that running 'next build' does not not respect the publicPath property in the webpack config (next.config.js).

I have opened a bug report with next on this here: [https://github.com/vercel/next.js/issues/40558](https://github.com/vercel/next.js/issues/40558) and it's only an issue for me in production because of the way that I structure my files on aws s3. Regardless of what I am doing with it though, webpack should prepend the publicPath to script path's for example but it does not (even when I run `next build` locally.
